### ISSUE TITLE
`possibly-unresolved-reference` false positive for walrus in `or`

### ISSUE THREAD

**Issue by Andrej730** [astral-sh/ty#4025]
### Summary

Consider case below - `c` will be assigned regardless whether `A` or `B` wins in `A or B`, but ty fails to detect it. 

```python
def test(a: bool):
    if ((a and (c:=54)) or (c:=32)):
        # Name `c` used when possibly not defined
        print(c)
```
[Playground](https://play.ty.dev/5fbb8df8-9c5c-42ee-8588-6398b7951b2d)
[pyright-play](https://pyright-play.net/?code=CYUwZgBALiDOUAoCGAuCAjA9pgNgShQCgISIBLSBZCJAO2AgQGMUBeAVgBY88JMAnRi1YBmAEw8ipaRADEEAHJIAtiAgADJuogBXWCAYB3ABYhaEAA6ZYsMuhwBPCLUxQIoMGVoHiMkhf4vRCY8IA) (pyright detect that variable is always defined)

PS Glad to see  `possibly-unresolved-reference` rule maturing now, I've tried it a few months ago and it was giving me too many false positives (most of them probably were related to walrus assignment IIRC), but now I went through 250+ reports from this rule and the only case that broke is this one. Thank you for the work!

### Version

ty 0.0.59

---

Can confirm as fixed in 0.0.61

**Comment by sharkdp** [comment]
Thank you for reporting this.

I'm afraid this might be hard to solve, unfortunately. It would require us to detect/understand this "xor" pattern.

**Comment by carljm** [comment]
@sharkdp It looks like it is possible to fix in a general way by adjusting the tracking of snapshots in boolean control flow; see attached PR.
